packages/sdk/observability/src/otel/metrics.ts

```diff
@@ -0,0 +1,114 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import {
+  type Attributes,
+  diag,
+  DiagConsoleLogger,
+  DiagLogLevel,
+  type Meter,
+  type ObservableGauge,
+} from '@opentelemetry/api';
+import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
+import { Resource } from '@opentelemetry/resources';
+import { MeterProvider, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
+import { SEMRESATTRS_SERVICE_NAME, SEMRESATTRS_SERVICE_VERSION } from '@opentelemetry/semantic-conventions';
+
+import { TRACE_PROCESSOR, type MetricData } from '@dxos/tracing';
+
+import { type OtelOptions } from './otel';
+
+const EXPORT_INTERVAL = 60 * 1000;
+const METRIC_PREFIX = 'dxos.';
+
+type SynchronousGauge = {
+  gauge: ObservableGauge<Attributes>;
+  nextValue: number;
+  nextTags?: any;
+};
+
+export class OtelMetrics {
+  private _meterProvider: MeterProvider;
+  private _meter: Meter;
+  private _gauges = new Map<string, SynchronousGauge>();
+
+  constructor(private readonly options: OtelOptions) {
+    // TODO: improve error handling/logging
+    //  https://github.com/open-telemetry/opentelemetry-js/issues/4823
+    const resource = Resource.default().merge(
+      new Resource({
+        [SEMRESATTRS_SERVICE_NAME]: this.options.serviceName,
+        [SEMRESATTRS_SERVICE_VERSION]: this.options.serviceVersion,
+      }),
+    );
+
+    const grafanaMetricReader = new PeriodicExportingMetricReader({
+      exporter: new OTLPMetricExporter({
+        url: this.options.endpoint + '/v1/metrics',
+        headers: {
+          Authorization: this.options.authorizationHeader,
+        },
+      }),
+      exportIntervalMillis: EXPORT_INTERVAL,
+    });
+
+    this._meterProvider = new MeterProvider({
+      resource,
+      readers: [grafanaMetricReader],
+    });
+    this._meter = this._meterProvider.getMeter('dxos-observability');
+
+    const metrics = {
+      // TODO: update metrics names and remove prefix?
```

Comment:
@wittjosiah @yarolegovich could we prefix the metric names (e.g. `dxos.echo.space.count`) so they're less ambiguous and consistent with the trace names?